File: cpp/epi/count_bits.cc
#include "generic_test.h"

short CountBits(unsigned int x) {
  short c = 0;

  while (x != 0) {
    ++c;
    x &= (x - 1);
  }

  return c;
}

short CountBitsCached(unsigned int x) {
  short a[256];
  for (unsigned int i = 0; i < 256; ++i) {
    a[i] = CountBits(i);
  }

  return a[(x >> 24) & 0xFF] +
          a[(x >> 16) & 0xFF] +
          a[(x >> 8) & 0xFF] +
          a[x & 0xFF];
}



int main(int argc, char* argv[]) {
  std::string func_name = "CountBits";
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"x"};

  GenericTestMain(args, "count_bits.cc", "../test_data/epi/count_bits.tsv",
                      &CountBits, DefaultComparator{}, func_name, param_names);
}

File: java/epi/CountBits.java
package epi;

import test_framework.EpiTest;
import test_framework.GenericTest;

public class CountBits {

  @EpiTest(testDataFile = "../test_data/epi/count_bits.tsv")
  public static short countBits(int x) {
    return 0;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "CountBits.java",
                new Object() {
                }.getClass().getEnclosingClass())
            .ordinal());
  }
}
